README.md
[![Build Status](https://travis-ci.org/bhofmei/jbplugin-smallrna.svg?branch=master)](https://travis-ci.org/bhofmei/jbplugin-smallrna)

# Small RNA Plugin

JBrowse plugin to support small RNA visualization

Based on the JBrowse tack type "Alignments2", the small RNA alignments track shows individual reads. The differences are:

1.  Reads are colored by size not strand  
2.  Reads can be filtered by size and/or multimapping  
3.  There is animal and plant specific coloring because plant's don't have piRNAs. Plant coloring is the default.
4.  Reads are organized on a y-axis by strand with positive-strand reads above the y-axis origin and negative-strand reads below it.


## Install

For JBrowse 1.11.6+ in the _JBrowse/plugins_ folder, type:  
``git clone https://github.com/bhofmei/jbplugin-smallrna.git SmallRNAPlugin``

**or**

downloaded the latest release version at [releases](https://github.com/bhofmei/jbplugin-smallrna/releases).  
Unzip the downloaded folder, place in _JBrowse/plugins_, and rename the folder _SmallRNAPlugin_

## Activate
Add this to _jbrowse.conf_ under `[GENERAL]`:

    [ plugins.SmallRNAPlugin ]
    location = plugins/SmallRNAPlugin

If that doesn't work, add this to _jbrowse_conf_.json:

    "plugins" : {
        "SmallRNAPlugin" : { "location" : "plugins/SmallRNAPlugin" }
    }
    
## Test

Sample data is included in the plugin to test that the plugin is working properly. With `URL` as the URL path to the JBrowse instance, navigate a web browser to `URL/index.html?data=plugins/SmallRNAPlugin/test/data`.

![Demo Image](img/demo_image.png)

## The Basics
- Each read is colored based on length
  - Blue: 21 nt
  - Green: 22 nt
  - Purple: 23 nt
  - Orange: 24 nt
  - Red: piRNAs (26-31 nt)
  - Gray: all other sizes
  - *Colors were inspired by [this color scheme](https://mpss.danforthcenter.org/web/php/pages/legend.php?SITE=AT_sRNA)*
- Reads are positions positions based on strand
  - Above y-axis origin: positive strand
  - Below y-axis origin: negative strand
- Unfilled reads indicate read maps to multiple locations
- Filtering options
  - Filter an individual track using the track drop-down menu
  - Filter all visible smRNA tracks using the toolbar button
  - Filter for size, strand, read quality, multimapping

## Plants vs Animals
The plant-specific smRNA color scheme is the default. However, if you want to use the animal-specific color scheme, it is very easy and flexible to change.  This is compatible with the [DNA Methylation plugin](https://github.com/bhofmei/jbplugin-methylation).
  
Using the animal coloring scheme is enforced hierarchically. Configurations specified at a higher level overpower lower-level specification. If not specified at a specific level, inherits the setting of the level below. 

| level| location | syntax|
|--|--|--|
|*highest* | individual track config | `isAnimal=true` |
| | `tracks.conf` | `[general]`<br>`isAnimal=true` |
| | `jbrowse.conf` | `[general]` <br> `isAnimal=true` |
|*lowest*| **default** | `isAnimal=false`|

Note that toolbar buttons are defined by `tracks.conf` and `jbrowse.conf`.

## Using small RNA Tracks
### File Formats
Reads are read directly from a BAM file. Follow the instructions for the Alignments2 track to specify the file location.  
Also, make sure the indexed BAM file (`.bam.bai`) is in the same directory as the BAM file.
Optionally, you can also supply a BigWig file for coverage view when zoomed out. I recommend this. Otherwise the track says "Too much data to show; zoom in to see detail"

### JSON Track Specifications
Track specifications are similar to the Alignments2 specifications. The _label_, _type_, and _urlTemplate_ must be specified. Take note of _type_; this is the difference from Alignments2.

To change the plant or animal coloring scheme, for the specific track, include the _isAnimal_ setting. Otherwise it assumes the setting based on the priority list above.

In trackList.json,

```
{
    "key" : "Small RNA",
    "label" : "smrna",
    "storeClass" : "JBrowse/Store/SeqFeature/BAM",
    "urlTemplate" : "path/to/smallrna.bam",
    "type" : "SmallRNAPlugin/View/Track/smAlignments",
    "isAnimal" : true
}
```

In tracks.conf,

    [tracks.smrna]
    key = Small RNA
    storeClass = JBrowse/Store/SeqFeature/BAM
    urlTemplate = path/to/smallrna.bam
    type = SmallRNAPlugin/View/Track/smAlignments
    isAnimal = true

Similar to other alignments, you can specify a BigWig file to use for the histogram view.

In trackList.json,

```
{
    "key" : "Small RNA",
    "label" : "smrna",
    "storeClass" : "JBrowse/Store/SeqFeature/BAM",
    "urlTemplate" : "path/to/smallrna.bam",
    "type" : "SmallRNAPlugin/View/Track/smAlignments",
    "isAnimal" : true,
    "histograms" : {
        "storeClass" : "JBrowse/Store/SeqFeature/BigWig",
        "urlTemplate" : "path/to/smallrna.bw"
    }
}
```

### HTML Features-style Track
The track type `smAlignments` is preferred for its speed, but HTMLFeature alignments track is available. 
It is comparable to JBrowse's Alignments track.

This track type is beneficial when taking PDF/SVG screenshots so each read is an HTML element (compared to being painted on the canvas for smAlignments/Alignments2).

Track configurations are the same except the track type and it does not accept a BigWig file for histograms.

In _tracks.conf_,
```
[tracks.smrna-html]
key = Small RNA - HTML
storeClass = JBrowse/Store/SeqFeature/BAM
urlTemplate = path/to/smallrna.bam
type = SmallRNAPlugin/View/Track/smHTMLAlignments
```

In _trackList.json_,
```
{
  "key" : "Small RNA - HTML",
  "label" : "smrna-html",
  "storeClass" : "JBrowse/Store/SeqFeature/BAM",
  "urlTemplate" : "path/to/smallrna.bam",
  "type" : "SmallRNAPlugin/View/Track/smHTMLAlignments",
  "isAnimal" : true
}
```

### Additional configurations
By default, mutlimapped reads are lighter/less opaque. To disable this, use `style.solidFill = true` in _tracks.conf_ and `"style":{ "solidFill" : true }` in _trackList.json_.

## Multimapping
Traditionally, for small RNA analysis is important to know if a read maps uniquely or multiple times within the genome. In this plugin, multi-mapped reads are shown in the appropriate color for the length, but are more transparent.  
Reads are determined to be multi mapped based on the SAM flag (read is supplementary) or the SAM attribute `NH`. Even when multiple alignments for the same read are output in the BAM file, the mapping program may not set the flag or `NH` attribute. For example, bowtie does not set these even when `-k 2+`.

## Quality Filtering
Mapping programs vary greatly in how they assign quality scores for reads, the `MQ` or `MAPQ` score. 
From the [SAM format specification](http://samtools.github.io/hts-specs/SAMv1.pdf),
  > MAPQ: MAPping Quality. It equals −10 log10 Pr{mapping position is wrong}, rounded to the nearest
integer. A value 255 indicates that the mapping quality is not available.

To accommodate as many scoring schemes as possible, users can filter by minimum quality score.

Reads with value 255 are always shown.

## Future Plans
The following features plan to be supported in the future.
- Testing for smHTMLAlignments

CHANGELOG.md
# Release Notes

## [v1.4.6] - 2019-04-08
- FIXED error where base and quality didn't show up properly in the alignment
dialog in the newer versions of JBrowse; caused by change in function parameters
in DetailsMixin function

##[v1.4.5] - 2019-03-25
- FIXED error with strand undefined in newer JBrowse versions; this was caused by a change in how JBrowse stores BAM data

## [v1.4.4] - 2018-04-27
- FIXED minor bug where it wouldn't work when using the NeatCanvasFeature plugin and there were multimapped reads

## [v1.4.3] - 2017-11-02
- UPDATED small rna colors
  - 21: blue
  - 22: green
  - 23: purple
  - 24: orange
  - piRNAs: red
  - other: gray
- FIXED minor bug with opacity in details dialog

## [v1.4.2] - 2017-10-02
- ADDED dialog mode to take screenshots of filtering dialog

## [v1.4.1] - 2017-08-17
- minor bug fixes

## [v1.4.0] - 2017-08-08
- ADDED an HTMLFeature style track to compliment the default canvas style track
- this is mainly useful for screenshots so reads are div objects not painted on the canvas

## [v1.3.1] - 2016-10-04
- FIXED issue with filtering when changing between multiple non-zero values
- FIXED some labels to be "mapping quality" not "quality score" to minimize possible confusion

## [v1.3.0] - 2016-09-26
- ADDED filtering option to filter by minimum quality score of reads
- Able to filter per track or across all tracks

## [v1.2.3] - 2016-09-24
- UPDATED Handle isAnimal better to match MethylationPlugin

## [v1.2.2] - 2016-09-23
- ADDED Option to solid-fill multimappers which will be useful for screenshots if user does not want multimappers to appear different

## [v1.2.1] - 2016-09-19
- FIXED issue with track check boxes IDs that prevented filtering on multiple tracks

## [v1.2.0] - 2016-06-10
- Reads are organized on the y-axis by strand; positive reads on positive-strand y-axis and negative-strand reads on negative y-axis
- Also includes "max height exceeded" warning independently for the positive and negative y-axis
- In order to accommodate more data, the individual reads are smaller than the default heights for RNA-seq reads

## [v1.1.0] - 2016-06-02
- ADDED Filtering dialog for all visible tracks
  - Only works on visible tracks and only changes features in which the check box was clicked
  - Colored checkboxes

## [v1.0.4] - 2016-04-03
- FIXED issue with coloring of multi mapped reads; now show transparency when multimapped

## [v1.0.3] - 2016-03-25
- FIXED Correctly display multimapping status in the read detail pop-up

## [v1.0.2] - 2016-03-18
- ADDED Use XM attribute (with sam flag and/or NH attribute) to determine if a read is multi mapped

## [v1.0.1] - 2016-03-14
- UPDATED coloring

## [v1.0.0] - 2016-03-09
- First production ready version
- Supports coloring and filtering of reads by size
- Supports animal vs non-animal smRNA scheme (animal includes piRNAs which are 26-31 bp)
- Supports coloring and filtering based on multimapped status


js/View/Track/_AlignmentsMixin.js
/**
 * Based on JBrowse _AlignmentsMixin
 * customized for smallRNA data
 * specifically, fewer track menu options, different colors, less details
 */
define("SmallRNAPlugin/View/Track/_AlignmentsMixin", [
           'dojo/_base/declare',
           'dojo/_base/array',
           'dojo/_base/lang',
           'dojo/when',
           'JBrowse/Util',
           'JBrowse/Store/SeqFeature/_MismatchesMixin',
           'SmallRNAPlugin/View/Track/_NamedFeatureFiltersMixin'
        ],
        function(
            declare,
            array,
            lang,
            when,
            Util,
            MismatchesMixin,
            NamedFeatureFiltersMixin
        ) {

return declare([ MismatchesMixin, NamedFeatureFiltersMixin ], {

    /**
     * Make a default feature detail page for the given feature.
     * @returns {HTMLElement} feature detail page HTML
     */
    defaultFeatureDetail: function( /** JBrowse.Track */ track, /** Object */ f, /** HTMLElement */ div ) {
        var container = dojo.create('div', {
            className: 'detail feature-detail feature-detail-'+track.name.replace(/\s+/g,'_').toLowerCase(),
            innerHTML: ''
        });
        var fmt = lang.hitch( this, function( name, value, feature, unsafe ) {
          if(unsafe===undefined) unsafe=false;
            name = Util.ucFirst( name.replace(/_/g,' ') );
            return this.renderDetailField(container, name, value, feature, null, {}, unsafe);
        });
        fmt( 'Name', f.get('name'), f );
        fmt( 'Type', f.get('type'), f );
        // if score is undefined, that means it was originally 255
        var score = (f.get('score'));
        fmt( 'Score', (score === undefined ? 255 : score), f );
        fmt( 'Description', f.get('note'), f );
        fmt(
            'Position',
            Util.assembleLocString({ start: f.get('start'),
                                     end: f.get('end'),
                                     ref: this.refSeq.name })
            + ({'1':' (+)', '-1': ' (-)', 0: ' (no strand)' }[f.get('strand')] || ''),
            f
        );


        if( f.get('seq') ) {
            fmt('Sequence and Quality', this._renderSeqQual( f ), f, true );
        }
        /* multimapping */
        var mm = (f.get('supplementary_alignment') || (typeof f.get('xm')!='undefined'&&f.get('xm')>1) || (typeof f.get('nh') != 'undefined' && f.get('nh') > 1 ));
        fmt("Multimapped",mm,f);

        /* change filtering options to only capture the options we are interested in
        seq, qual, supplementary_alignment, source, seq_length, NH, CIGAR, seq_reverse_complemented(?)*/
        var additionalTags = array.filter(
            f.tags(), function(t) {
                return {source:1,seq_length:1,NH:1,CIGAR:1,seq_reverse_complemented:1}[t.toLowerCase()];
            }
        );
        array.forEach( additionalTags, function(t) {
            if(t=="NH")
                fmt("Number mapped locations",f.get(t),f);
            else
                fmt( t, f.get(t), f );
        });

        return container;
    },

    // takes a feature, returns an HTML representation of its 'seq'
    // and 'qual', if it has at least a seq. empty string otherwise.
    _renderSeqQual: function( feature ) {

        var seq  = feature.get('seq'),
            qual = feature.get('qual') || '';
        if( !seq )
            return '';

        qual = qual.split(/\s+/);

        var html = '';
        for( var i = 0; i < seq.length; i++ ) {
            html += '<div class="basePosition" title="position '+(i+1)+'"><span class="seq">'
                    + seq[i]+'</span>';
            if( qual[i] )
                html += '<span class="qual">'+qual[i]+'</span>';
            html += '</div>';
        }
        return '<div class="baseQuality">'+html+'</div>';
    },

    // recursively find all the stylesheets that are loaded in the
    // current browsing session, traversing imports and such
    _getStyleSheets: function( inSheets ) {
        var outSheets = [];
        array.forEach( inSheets, function( sheet ) {
            outSheets.push( sheet );
            array.forEach( sheet.cssRules || sheet.rules, function( rule ) {
                if( rule.styleSheet )
                    outSheets.push.apply( outSheets, this._getStyleSheets( [rule.styleSheet] ) );
            },this);
        },this);
        return outSheets;
    },

    // filters for BAM alignments according to some flags
    /* this function needs updated */
    _getNamedFeatureFilters: function() {
        return lang.mixin( {}, this.inherited( arguments ),
        {
            hideMultiMappers: {
                desc: 'Hide multi-mapped alignments',
                title: 'Show only uniquely aligned reads',
                func: function( f ) {
                    return ! (f.get('supplementary_alignment') || (typeof f.get('nh') != 'undefined' && f.get('nh') > 1 ) || (typeof f.get('xm')!='undefined'&&f.get('xm')>1));
                }
            },
            filterQuality: {
                desc: 'Filter by mapping quality',
                title: 'Set minimum mapping quality for visible reads',
                alwaysfilter: true,
                func: function( f ){
                    return ( (this.config.filterQuality === 0) || ((f.get('score') === undefined) || f.get('score') === 255 || f.get('score') >= this.config.filterQuality ))
                }
            },
            hideForwardStrand: {
                desc: 'Hide reads aligned to the forward strand',
                title:'Show/hide forward strand reads',
                func: function( f ) {
                    return f.get('strand') != 1;
                }
            },
            hideReverseStrand: {
                desc: 'Hide reads aligned to the reverse strand',
                title:'Show/hide reverse strand reads',
                func: function( f ) {
                    return f.get('strand') != -1;
                }
            },
            hide21:{
                desc: 'Hide 21-mers',
                title: 'Show/hide 21 bp-long reads',
                id: 'smrna-select-blue',
                func: function(f){
                    return f.get('seq_length') != 21;
                }
            },
            hide22:{
                desc: 'Hide 22-mers',
                title: 'Show/hide 22 bp-long reads',
                id: 'smrna-select-green',
                func: function(f){
                    return f.get('seq_length') != 22;
                }
            },
            hide23:{
                desc: 'Hide 23-mers',
                title: 'Show/hide 23 bp-long reads',
                id: 'smrna-select-purple',
                func: function(f){
                    return f.get('seq_length') != 23;
                }
            },
            hide24:{
                desc: 'Hide 24-mers',
                title: 'Show/hide 24 bp-long reads',
                id: 'smrna-select-orange',
                func: function(f){
                    return f.get('seq_length') != 24;
                }
            },
            hidepi:{
                desc: 'Hide piRNAs',
                title: 'Show/hide piRNAs (26-31 bp)',
                id: 'smrna-select-red',
                func: function(f){
                    return !(f.get('seq_length') > 25 && f.get('seq_length') < 32);
                }
            },
            hideOthers:{
                desc: 'Hide others',
                title: 'Show/hide all other sized reads',
                id: 'smrna-select-gray',
                func: function(f){
                    return this.config.isAnimal ? !(f.get('seq_length') < 21 || f.get('seq_length') > 31 || f.get('seq_length')==25) : !(f.get('seq_length') < 21 || f.get('seq_length') > 24);
                }
            }
        });
    },

    _alignmentsFilterTrackMenuOptions: function() {
        // add toggles for feature filters
        var track = this;
        return when( this._getNamedFeatureFilters() )
            .then( function( filters ) {
                var sizesAr = ['hide21','hide22','hide23','hide24'];
                if (track.config.isAnimal)
                    sizesAr.push('hidepi');
                sizesAr.push('hideOthers');
                return track._makeFeatureFilterTrackMenuItems(
                   [
                       'filterQuality',
                       'hideForwardStrand',
                       'hideReverseStrand',
                       'hideMultiMappers'
                   ],
                    sizesAr,
                   filters );
            });
    }

});
});
